src/olympia/scanners/tests/test_admin.py

```diff
@@ -601,3 +609,214 @@ def test_get_fields_for_non_admins(self):
+        url = reverse(
+            'admin:scanners_scannerqueryrule_handle_abort', args=(rule.pk, ))
+        button = field.find('button')[0]
+        assert button.attrib['formaction'] == url
+
+    def test_no_button_for_completed_rule_query(self):
+        ScannerQueryRule.objects.create(
+            name='bar', scanner=YARA, state=COMPLETED)
+        response = self.client.get(self.list_url)
+        assert response.status_code == 200
+        doc = pq(response.content)
+        field = doc('.field-state_with_actions')
+        assert field
+        assert field.text() == 'Completed'
+        assert not field.find('button')
+
+    def test_button_in_change_view(self):
+        rule = ScannerQueryRule.objects.create(
+            name='bar', scanner=YARA, state=RUNNING)
+        change_url = reverse(
+            'admin:scanners_scannerqueryrule_change', args=(rule.pk,))
+        response = self.client.get(change_url)
+        assert response.status_code == 200
+        doc = pq(response.content)
+        field = doc('.field-state_with_actions')
+        assert field
+        assert field.text() == 'State:\nRunning \xa0 Abort'
+        url = reverse(
+            'admin:scanners_scannerqueryrule_handle_abort', args=(rule.pk, ))
+        button = field.find('button')[0]
+        assert button.attrib['formaction'] == url
+
+    @mock.patch('olympia.scanners.admin.run_yara_query_rule.delay')
+    def test_run_action(self, run_yara_query_rule_mock):
+        rule = ScannerQueryRule.objects.create(
+            name='bar', scanner=YARA, state=NEW)
+        response = self.client.post(
+            reverse(
+                'admin:scanners_scannerqueryrule_handle_run',
+                args=[rule.pk],
+            ), follow=True
+        )
+        assert response.status_code == 200
+        assert response.redirect_chain == [(self.list_url, 302)]
+        assert run_yara_query_rule_mock.call_count == 1
+        assert run_yara_query_rule_mock.call_args[0] == (rule.pk,)
+        messages = list(response.context['messages'])
+        assert len(messages) == 1
+        assert 'successfully queued' in str(messages[0])
+        rule.reload()
+        assert rule.state == RUNNING
+
+    @mock.patch('olympia.scanners.admin.run_yara_query_rule.delay')
+    def test_run_action_wrong_state(self, run_yara_query_rule_mock):
+        rule = ScannerQueryRule.objects.create(
+            name='bar', scanner=YARA, state=ABORTING)
+        response = self.client.post(
+            reverse(
+                'admin:scanners_scannerqueryrule_handle_run',
+                args=[rule.pk],
+            ), follow=True
+        )
+        assert response.status_code == 200
+        assert response.redirect_chain == [(self.list_url, 302)]
+        assert run_yara_query_rule_mock.call_count == 0
+        messages = list(response.context['messages'])
+        assert len(messages) == 1
+        assert 'could not be queued' in str(messages[0])
```

Comment:
Should you also assert that the rule's `pk` and `state` appear in the message?

Reply:
Yeah, I was lazy. I'll do that.